exporter/exportertest/mock_receiver.go

```diff
@@ -0,0 +1,221 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package exportertest // import "go.opentelemetry.io/collector/exporter/exportertest"
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+var errNonPermanent = status.Error(codes.DeadlineExceeded, "non Permanent error")
+var errPermanent = status.Error(codes.Internal, "Permanent error")
+
+type MockReceiver interface {
+	Start() error
+	Stop() error
+}
+
+type MockReceiverFactory func(consumer *MockConsumer) MockReceiver
+
+// // randomNonPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// // half of the time and fails with a non-permanent error the rest of the time.
+func randomNonPermanentErrorConsumeDecision() error {
+	if rand.Float32() < 0.5 {
+		return errNonPermanent
+	}
+	return nil
+}
+
+// randomPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// half of the time and fails with a permanent error the rest of the time.
+func randomPermanentErrorConsumeDecision() error {
+	if rand.Float32() < 0.5 {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	return nil
+}
+
+// randomErrorsConsumeDecision is a decision function that succeeds approximately
+// a third of the time, fails with a permanent error the third of the time and fails with
+// a non-permanent error the rest of the time.
+func randomErrorsConsumeDecision() error {
+	r := rand.Float64()
+	third := 1.0 / 3.0
+	if r < third {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	if r < 2*third {
+		return errNonPermanent
+	}
+	return nil
+}
+
+type MockConsumer struct {
+	reqCounter          requestCounter
+	mux                 sync.Mutex
+	exportErrorFunction func() error
+	ReceivedTraces      []ptrace.Traces
+	ReceivedMetrics     []pmetric.Metrics
+	ReceivedLogs        []plog.Logs
+}
+
+func NewMockConsumer(decisionFunc func() error) MockConsumer {
+	return MockConsumer{
+		reqCounter:          newRequestCounter(),
+		mux:                 sync.Mutex{},
+		exportErrorFunction: decisionFunc,
+		ReceivedTraces:      nil,
+		ReceivedMetrics:     nil,
+		ReceivedLogs:        nil,
+	}
+}
+
+type requestCounter struct {
+	success int
+	error   errorCounter
+	total   int
+}
+
+type errorCounter struct {
+	permanent    int
+	nonpermanent int
+}
+
+func newErrorCounter() errorCounter {
+	return errorCounter{
+		permanent:    0,
+		nonpermanent: 0,
+	}
+}
+
+func newRequestCounter() requestCounter {
+	return requestCounter{
+		success: 0,
+		error:   newErrorCounter(),
+		total:   0,
+	}
+}
```

Comment:
Please move it after the `MockConsumer` methods

Reply:
done